sui-execution/latest/sui-move-natives/src/dynamic_field.rs

```diff
@@ -231,6 +232,13 @@ pub fn add_child_object(
             * struct_tag_size.into()
     );
 
+    if object_is_shared(context, &child)? {
+        return Ok(NativeResult::err(
+            context.gas_used(),
+            E_SHARED_OBJECT_OPERATION_NOT_SUPPORTED,
+        ));
+    }
+
```

Comment:
Similar to the `ObjectRuntime::transfer` case, it feels reasonable to me that `add_child_object` would itself and report if the operation is allowed or not. 

Though, I'll admit it feels less clean than the normal `transfer` case